lib/SIL/SILFunctionType.cpp

```diff
@@ -1059,11 +1059,122 @@ static CanSILFunctionType getSILFunctionType(
     .withIsPseudogeneric(pseudogeneric)
     .withNoEscape(extInfo.isNoEscape());
   
+  // Extract the abstract generic calling convention of the function into a
+  // substituted generic signature for the function type.
+  //
+  // This signature only needs to consider the general calling convention,
+  // so it can reduce away protocol and base class constraints aside from
+  // `AnyObject`. We wanto similar-shaped generic function types to remain
+  // canonically equivalent, like `(T, U) -> ()`, `(T, T) -> ()`,
+  // `(U, T) -> ()` or `(T, T.A) -> ()` when given substitutions that produce
+  // the same function types, so we also introduce a new generic argument for
+  // each position where we see a dependent type, and canonicalize the order in
+  // which we see independent generic arguments.
+  //
+  bool impliedSignature = false;
+  SubstitutionMap substitutions;
+  if (TC.Context.LangOpts.EnableSubstSILFunctionTypesForFunctionValues
+      // We don't currently use substituted function types for generic function
+      // type lowering, though we should for generic methods on classes and
+      // protocols.
+      && !genericSig) {
+    class SubstFunctionTypeCollector {
+    public:
+      TypeConverter &TC;
+      const bool mapReplacementsOutOfContext;
+      
+      SmallVector<GenericTypeParamType *, 4> substGenericParams;
+      SmallVector<Requirement, 4> substRequirements;
+      SmallVector<Type, 4> substReplacements;
+      
+      SubstFunctionTypeCollector(TypeConverter &TC,
+                                 bool mapReplacementsOutOfContext)
+        : TC(TC), mapReplacementsOutOfContext(mapReplacementsOutOfContext) {}
+      SubstFunctionTypeCollector(const SubstFunctionTypeCollector &) = delete;
+
+      // TypeSubstitutionFn
+      Type operator()(SubstitutableType *t) {
+        ArchetypeType *archetype = dyn_cast<ArchetypeType>(t);
+        if (!archetype)
+          archetype = TC.getCurGenericContext()->getGenericEnvironment()
+                                   ->mapTypeIntoContext(t)
+                                   ->castTo<ArchetypeType>();
+        
+        // Replace every dependent type we see with a fresh type variable in
+        // the substituted signature.
+        auto paramIndex = substGenericParams.size();
+        auto param = CanGenericTypeParamType::get(0, paramIndex, TC.Context);
+
+        substGenericParams.push_back(param);
+        Type replacementTy = t;
+        if (mapReplacementsOutOfContext) {
+          replacementTy = t->mapTypeOutOfContext();
+        }
+        substReplacements.push_back(replacementTy);
+        
+        // Preserve the AnyObject constraint, if any, on the archetype in the
+        // generic signature.
+        if (archetype->requiresClass()) {
+          substRequirements.push_back(Requirement(RequirementKind::Layout,param,
+           LayoutConstraint::getLayoutConstraint(LayoutConstraintKind::Class)));
+        }
```

Comment:
Well, some of the other layout constraints definitely seem useful to distinguish ABI-wise, and I'm not sure there's harm in just preserving all the original layout constraints.  If we specialized an implementation, presumably function arguments would (by default) preserve the original constraints of the function, and it would just be the substitution with the specialized constraint.

Reply:
The main issue I see with that is that base class constraints imply `_NativeClass` instead of `AnyObject` as their layout constraint when the base class has no ObjC ancestry, and that seems like it could introduce a lot of spurious type mismatches.

Reply:
Oh, I hadn't thought of that.  Yeah, maybe doing it selectively would make more sense.